File: sequential/nilikantha.cpp
/*----------------------------------------------------------------
* Multiprocesadores: Proyecto final
* Fecha: 21-Nov-2021
* Autor: A01654856 Hugo David Franco Ávila
* Descripción: Este código implementa la serie de Nilikantha para
    obtener una aproximación suficientemente precisa de Pi en el lenguaje C++. A medida
    que incrementan los términos de la serie, la aproximación es más precisa.
    El algoritmo está de forma secuencial.
    Hago uso del archivo utils.h generado por el Prof. Pedro Pérez
* Comando para compilar en Linux: g++ nilikantha.cpp
* Comando para correr en Linux: ./a.out
* NOTA: De especificar otro nombre del ejecutable al compilar utilizar ./nombreDelEjecutable.out
*--------------------------------------------------------------*/
#include <iostream>
#include <iomanip>
#include "utils.h"

#define SIZE 1000000000 //1e9

using namespace std;

class Nilikantha {
private:
    int size;

    double result;

public:
    /**
     * @brief Construct a new Nilikantha object
     * 
     * @param s int
     */
    Nilikantha(int s) : size(s) {}

    /**
     * @brief Get the result value from the Nilikantha object
     * 
     * @return double 
     */
    double getResult() {
        return this->result;
    }

    /**
     * @brief This function does the calculation of the Nilikantha series up to the size defined
     * in the object constructor.
     * 
     */
    void calculate() {
        result = 3.0;
        double sign, di, denominator, term;
        int i = 1;
        for (i; i <= size; i++) {
            sign = i % 2 == 0 ? -1.0 : 1.0;
            di = i * 2.0;
            denominator = (di * (di + 1) * (di + 2));
            term = (4.0 * sign) / denominator;
            result += term;
        }
    }
};

int main(int argc, char *argv[]) {
    int size = SIZE, i = 0;
    double result = 0.0, time = 0.0;
    if (argc >= 2) {
        cout << "Error: No arguments are allowed" << endl;
        return -1;
    }
    cout << "Starting..." << endl;
    Nilikantha nik(size);
    for (i = 0; i < N; i++) {
        start_timer();
        
        nik.calculate();

        time += stop_timer();
    }
    cout << "Pi is approx: " << setprecision(15) << nik.getResult() << endl;
    cout << "Average time elapsed: " << setprecision(15) << (time / N) << " ms" << endl;
    return 0;
}

File: sequential/Nilikantha.java
/*----------------------------------------------------------------
* Multiprocesadores: Proyecto final
* Fecha: 21-Nov-2021
* Autor: A01654856 Hugo David Franco Ávila
* Descripción: Este código implementa la serie de Nilikantha para
    obtener una aproximación suficientemente precisa de Pi en el lenguaje Java. A medida
    que incrementan los términos de la serie, la aproximación es más precisa.
    El algoritmo está de forma secuencial.
* Comando para compilar en Linux: javac Nilikantha.java
* Comando para correr en Linux: java Nilikantha
*--------------------------------------------------------------*/
public class Nilikantha {
    private static final int SIZE = 1_000_000_000;
    private static final int N = 10;
    private double result;

    /**
     * Default constructor for the Nilikantha class
     */
    public Nilikantha(){}

    /**
     * This function receives an integer s, and calculates the Nilikantha series up to the s-term,
     * which represents an approximation to the term Pi. The larger the value of S, the better 
     * approximation you  get.
     * 
     * @param s int
     */
    public void calculate(int s){
        result = 3.0;
        double sign, di, denominator, term;
        for(int i = 1; i <= s; i++){
            sign = i % 2 == 0 ? -1.0 : 1.0;
            di = i*2.0;
            denominator = (di*(di+1)*(di+2));
            term = (4.0*sign)/denominator;
            result += term;
        }
    }

    /**
     * This function returns the value of result in the Nilikantha object
     * @return double
     */
    private double getResult(){
        return this.result;
    }

    public static void main(String args[]){
        if(args.length >= 1){
            System.out.println("Error: No arguments are allowed");
            return;
        }
        long startTime, stopTime;
		double acum = 0;

		Nilikantha n = new Nilikantha();
		acum = 0;
		System.out.printf("Starting...\n");
		for (int i = 0; i < N; i++) {
			startTime = System.currentTimeMillis();

			n.calculate(SIZE);

			stopTime = System.currentTimeMillis();

			acum += (stopTime - startTime);
		}
		System.out.println("Pi is approx: " + n.getResult());
		System.out.println("Average time elapsed: " + (acum / N) + " ms");
    }
}